### PR TITLE
feat: implement time lease based locks for the `CryptoStore`

### PR DESCRIPTION
**Pull request by bnjbvr**
This implements a new time lease based lock for the `CryptoStore`, that doesn't require explicit unlocking, so that's more robust in the context of #1928, where any process may die because the device is running out of battery, or unexpected flows cause a lock to not be released properly in one or the other process.

```
//! This is a per-process lock that may be used only for very specific use
//! cases, where multiple processes might concurrently write to the same
//! database at the same time; this would invalidate crypto store caches, so
//! that should be done mindfully. Such a lock can be acquired multiple times by
//! the same process, and it remains active as long as there's at least one user
//! in a given process.
//!
//! The lock is implemented using time-based leases to values inserted in a
//! crypto store. The store maintains the lock identifier (key), who's the
//! current holder (value), and an expiration timestamp on the side; see also
//! `CryptoStore::try_take_leased_lock` for more details.
//!
//! The lock is initially acquired for a certain period of time (namely, the
//! duration of a lease, aka `LEASE_DURATION_MS`), and then a "heartbeat" task
//! renews the lease to extend its duration, every so often (namely, every
//! `EXTEND_LEASE_EVERY_MS`). Since the tokio scheduler might be busy, the
//! extension request should happen way more frequently than the duration of a
//! lease, in case a deadline is missed. The current values have been chosen to
//! reflect that, with a ratio of 1:10 as of 2023-06-23.
//!
//! Releasing the lock happens naturally, by not renewing a lease. It happens
//! automatically after the duration of the last lease, at most.
```

## Notes

- This is not implemented on indexeddb, because time in wasm is slightly more complicated (requires a host API). Maybe we already have that, but I haven't bothered since I'm not sure that wasm may need the lock in the first place (maybe for a Web app using the SDK, running from multiple tabs?).
- I've kept the former implementation for `insert_custom_value_if_missing`/`remove_custom_value` at the moment, but we could remove them if this lock proves to be more robust.

### CODE REVIEW

**Comment by ara4n** [comment]
why do we need time leased locks? when i tested this with file based posix locks, you could acquire the lock no matter how dirtily the process holding the lock got knifed.

**Comment by poljar** [comment]
> why do we need time leased locks? when i tested this with file based posix locks, you could acquire the lock no matter how dirtily the process holding the lock got knifed.

There isn't a great and obvious choice for a `flock` based crate and this was easier to implement in a platform-independent way in a timely manner. We certainly need to revise some parts of this as we go along.

File: crates/matrix-sdk-crypto/src/store/locks.rs [hunk 1]
```diff
@@ -14,21 +14,40 @@
...
+//! extension request should happen way more frequently than the duration of a
+//! lease, in case a deadline is missed. The current values have been chosen to
+//! reflect that, with a ratio of 1:10 as of 2023-06-23.
+//!
+//! Releasing the lock happens naturally, by not renewing a lease. It happens
+//! automatically after the duration of the last lease, at most.
+//!
+//! For this to work, the store must allow concurrent reads and writes from
+//! multiple processes. For instance, for sqlite, this means that it is running in [WAL](https://www.sqlite.org/wal.html)
+//! mode.
```
Comment (jplatte) [inline crates/matrix-sdk-crypto/src/store/locks.rs:42]: I don't understand this. Why is this necessary?

File: crates/matrix-sdk-crypto/src/store/locks.rs [hunk 2]
```diff
@@ -89,53 +137,95 @@ impl CryptoStoreLock {
...
     }
 
     /// Try to lock once, returns whether the lock was obtained or not.
     #[instrument(skip(self), fields(?self.lock_key, ?self.lock_holder))]
-    pub async fn try_lock_once(&self) -> Result<bool, CryptoStoreError> {
-        // Hold the num_holders lock for the entire's function lifetime, to avoid
-        // internal races if called in a reentrant manner.
-        let mut holders = self.num_holders.lock().await;
+    pub async fn try_lock_once(&self) -> Result<Option<CryptoStoreLockGuard>, CryptoStoreError> {
+        // Hold onto the locking attempt mutext for the entire lifetime of this
```
Comment (jplatte) [inline crates/matrix-sdk-crypto/src/store/locks.rs:148]:
```suggestion
        // Hold onto the locking attempt mutex for the entire lifetime of this
```

**Review by jplatte — COMMENTED** [review]
This is just to note that I've started review. I haven't gotten very far yet, but will continue tomorrow.

File: crates/matrix-sdk-crypto/src/store/locks.rs [hunk 3]
```diff
@@ -68,6 +106,16 @@ pub struct CryptoStoreLock {
 }
 
 impl CryptoStoreLock {
+    /// Amount of time a lease of the lock should last, in milliseconds.
+    pub const LEASE_DURATION_MS: u32 = 2000;
```
Comment (Hywan) [inline crates/matrix-sdk-crypto/src/store/locks.rs:110]: Could be 1000, or even 800, so that it cannot be really perceived by the app end-user?

File: crates/matrix-sdk-crypto/src/store/locks.rs [hunk 4]
```diff
@@ -68,6 +106,16 @@ pub struct CryptoStoreLock {
...
+
+    /// Period of time between two attempts to extend the lease. We'll
+    /// re-request a lease for an entire duration of `LEASE_DURATION_MS`
+    /// milliseconds, every `EXTEND_LEASE_EVERY_MS`, so this has to
+    /// be an amount safely low compared to `LEASE_DURATION_MS`, to make sure
+    /// that we can miss a deadline without compromising the lock.
+    const EXTEND_LEASE_EVERY_MS: u64 = 200;
```
Comment (Hywan) [inline crates/matrix-sdk-crypto/src/store/locks.rs:117]: Could be 50 or 100?

File: crates/matrix-sdk-crypto/src/store/locks.rs [hunk 5]
```diff
@@ -89,53 +137,95 @@ impl CryptoStoreLock {
...
-            return Ok(true);
-        }
+        if acquired {
+            // This is the first time we've acquired the lock. We're going to spawn the task
+            // that will renew the lease.
+
+            // Clone data to be owned by the task.
+            let this = self.clone();
+
+            matrix_sdk_common::executor::spawn(async move {
```
Comment (Hywan) [inline crates/matrix-sdk-crypto/src/store/locks.rs:177]: Are we sure we don't want to keep a handle to this spawned task, so that we can abort it in case of emergency?

**Review by Hywan — APPROVED** [review]
Overall it looks good to me :-).

File: crates/matrix-sdk-crypto/src/store/locks.rs (hunk 1 above)
Comment (poljar) [reply crates/matrix-sdk-crypto/src/store/locks.rs:42]: Yup, I don't think WAL mode is relevant here. You can still have multi-process support without WAL, but it'll block on the SQLite side.

File: crates/matrix-sdk/src/encryption/mod.rs [hunk 6]
```diff
@@ -876,46 +878,40 @@ impl Encryption {
...
-                if let Err(err) = self.client.base_client().regenerate_olm().await {
-                    // First, give back the cross-process lock.
-                    lock.unlock().await?;
-                    // Then return the error to the caller.
-                    return Err(err.into());
-                };
+            match lock.try_lock_once().await? {
+                Some(guard) => Ok(Some(guard)),
+                None => {
+                    // We didn't get the lock on the first attempt, so that means that another
```
Comment (poljar) [inline crates/matrix-sdk/src/encryption/mod.rs:889]: Are we certain that we don't need to reload when we did acquire the lock on the first attempt? Isn't there a race where we might get the lock just as it was released by the other process?

**Review by poljar — APPROVED** [review]
I think this looks good. It would be nice if people on the EX side would test this but I guess that's what the nightly is there for.

From my point of view, we can tweak the timeouts if they end up being too long later on.

File: crates/matrix-sdk-crypto/src/store/locks.rs (hunk 1 above)
Comment (bnjbvr) [reply crates/matrix-sdk-crypto/src/store/locks.rs:42]: Ah, cool; I misunderstood the role of WAL then. Removed that paragraph from the comment.

File: crates/matrix-sdk-crypto/src/store/locks.rs (hunk 2 above)
Comment (bnjbvr) [reply crates/matrix-sdk-crypto/src/store/locks.rs:148]: Thankts.

File: crates/matrix-sdk-crypto/src/store/locks.rs (hunk 3 above)
Comment (bnjbvr) [reply crates/matrix-sdk-crypto/src/store/locks.rs:110]: Fwiw: we chatted about it, and we'll try renew-every 50ms / lease-for 500ms: still a 1:10 ratio, and the lease should be closer to non-perceivable by the user (ideally we'd go under 300ms, but that seems more dangerous with respect to the scheduler being super busy).

File: crates/matrix-sdk/src/encryption/mod.rs (hunk 6 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/encryption/mod.rs:889]: Indeed you're right! This is going to be more generally fixed by #2155 :-)

File: crates/matrix-sdk-crypto/src/store/locks.rs (hunk 5 above)
Comment (bnjbvr) [reply crates/matrix-sdk-crypto/src/store/locks.rs:177]: Yes, it's also useful to make sure we don't spawn it multiple times :+1: Added some code to handle that.

File: crates/matrix-sdk-crypto/src/store/locks.rs [hunk 7]
```diff
@@ -177,57 +177,64 @@ impl CryptoStoreLock {
...
-                                break;
-                            }
-                        }
+            // Cancel the previous task, if any. That's safe to do, because:
+            // - either the task was done,
+            // - or it was still running, but taking a lock in the db has to be an atomic
+            //   operation
+            // running in a transaction.
+
+            drop(renew_task.take());
```
Comment (Hywan) [inline crates/matrix-sdk-crypto/src/store/locks.rs:186]: Be careful, it doesn't abort the task!

**Comment by Velin92** [comment]
Tested and this thing is bulletproof, tested on scenarios where both the NSE and Main App did not stop the sync and did not release the lock, and everything worked as expected.
@bnjbvr

File: crates/matrix-sdk-crypto/src/store/locks.rs (hunk 7 above)
Comment (bnjbvr) [reply crates/matrix-sdk-crypto/src/store/locks.rs]: Thanks, TIL. I've added an `abort()` that's guarded by `#[cfg(not(target_arch = "wasm32"))]`; this code wouldn't work on wasm32 anyways because it's using tokio::sleep.
